enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,216 @@
+        - quay.io/etcd.v1.0.0
+```
+
+A new opm sub-command, `represent` will be introduced that will output a yaml or a json representation of an index. 
+
+I.e, `opm represent --output=yaml docker.io/my-namespace/my-index` will output the `index.yaml` in the example above.
+
+The [API repo](https://github.com/operator-framework/api) will contain the `index.CUE` file that'll contain the definitions for an index. 
+
+An example CUE definition of an index can be: 
+
+```definitions.CUE
+import "encoding/json"
+
+#Constraint: {
+	type:  string
+    #serialized: string
+	...
+}
+
+#ConstraintSet: [...#Constraint]
+
+#PackageConstraint: #Constraint & {
+	type:        "olm.package"
+	packageName: string
+	// TODO: regex version ranges
+	version: string
+    #serialized: json.Marshal({"pkg": packageName, "version": version})
+}
+
+#LabelConstraint: #Constraint & {
+	type:  "label"
+	value: string
+    #serialized: value
+}
+
+#GVKConstraint: #Constraint & {
+	type: "olm.gvk"
+	// TODO: regexes
+	group:   string
+	version: string
+	kind:    string
+    #serialized: json.Marshal({"group": group, "version": version, "kind": kind})
+}
+
+#Provides: {
+    type: string
+    #serialized: string
+    ...
+}
+
+#PackageProvides: #Provides & {
+    type:        "olm.package"
+	packageName: string
+	// TODO: regex semver version
+	version: string
+    #serialized: json.Marshal({"pkg": packageName, "version": version})
+}
+
+#GVKProvides: #Provides & {
+    type: "olm.gvk"
+	// TODO: regexes
+	group:   string
+	version: string
+	kind:    string
+    #serialized: json.Marshal({"group": group, "version": version, "kind": kind})
+}
+
+#LabelProvides: #Provides & {
+	type:  "label"
+	value: string
+    #serialized: value
+}
+
+#Node: {
+	bundlePath:   string
+	provides?:    [...#Provides]
+	constraints?: [...#ConstraintSet]
+}
+
+#Edge: {
+    from: string,
+    to: string,
+    _bad: *false | true
+}
+
+#EdgeMap: [Name=string]: {
+    [To=string]: #Edge & {
+        from: Name
+        to: To
+    }
+}
+
+// edge is a list of graph edges between nodes
+_edge: #EdgeMap
+
+// nedge is a list of rejected graph edges that should not be followed
+_nedge: #EdgeMap
+
+graph: nodes: [...#Node]
+```
+
+With this definition, and index representation can be validated using CUE: `cue vet representation.CUE index.yaml`.
+
+A new opm sub-command `validate` will be introduced, that'll leverage CUE's vet command to provide the same validation functionality as CUE's `vet` command. 
+I.e to validate an index, the index author will use `opm index validate index.yaml representation.CUE`  
+
+> Note: A cue file hosted on github could possible be used in the `vet` command if [this issue](https://github.com/cuelang/cue/issues/505) is resolved. When this functionality is made available, the default CUE definition can be set to `https://github.com/operator-framework/api/index.CUE`, so that `opm index validate index.yaml` will be sufficient to validate and index.   
```

Comment:
Is this going to be a requirement? What about offline environments?

Reply:
The idea is that the hosted definition will be a `default`, which means if another CUE file is passed on to the vet command, it will override the default and opm will not try to reach out to an external environment.  

> What about offline environments?
Gotta admit I wasn't thinking about offline environments but I guess the nature of `default` solves that problem.